experimental/skills/DialogToDialog/DialogRootBot/SkillConversationIdFactory.cs

```diff
@@ -0,0 +1,75 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Bot.Builder;
+using Microsoft.Bot.Builder.Skills;
+using Microsoft.Bot.Schema;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.BotBuilderSamples.DialogRootBot
+{
+    /// <summary>
+    /// A <see cref="SkillConversationIdFactory"/> that uses <see cref="IStorage"/> to store and retrieve <see cref="ConversationReference"/> instances.
+    /// </summary>
+    public class SkillConversationIdFactory : SkillConversationIdFactoryBase
+    {
+        private readonly IStorage _storage;
+
+        public SkillConversationIdFactory(IStorage storage)
+        {
+            _storage = storage ?? throw new ArgumentNullException(nameof(storage));
+        }
+
+        public override async Task<string> CreateSkillConversationIdAsync(ConversationReference conversationReference, CancellationToken cancellationToken)
+        {
+            if (conversationReference == null)
+            {
+                throw new ArgumentNullException(nameof(conversationReference));
+            }
+
+            if (string.IsNullOrWhiteSpace(conversationReference.Conversation.Id))
+            {
+                throw new NullReferenceException($"ConversationId in {nameof(conversationReference)} can't be null.");
+            }
+
+            if (string.IsNullOrWhiteSpace(conversationReference.ChannelId))
+            {
+                throw new NullReferenceException($"ChannelId in {nameof(conversationReference)} can't be null.");
+            }
+
+            var storageKey = $"{conversationReference.Conversation.Id}{conversationReference.ChannelId}".GetHashCode().ToString(CultureInfo.InvariantCulture);
```

Comment:
This doesn't feel very robust, as the hash code could generate duplicate keys (in theory).

Reply:
Changed how the string is generated (the parent CID + the channel + plus an arbitrary string).

---
In reply to: [356240406](https://github.com/microsoft/BotBuilder-Samples/pull/2032#discussion_r356240406) [](ancestors = 356240406)